### PR TITLE
Add new operation for Duplicate Faces, which is similar to detach but…

### PR DESCRIPTION
**Pull request by MartinImgSpc**
… leaves the origin geometry unchanged.

### CODE REVIEW

**Comment by karl-** [comment]
image

**Comment by MartinImgSpc** [comment]
@karl- thanks I added the missing meta

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            if (count > 0)
+                return new ActionResult(ActionResult.Status.Success, "Duplicate " + count + (count > 1 ? " Faces" : " Face"));
+
+            return new ActionResult(ActionResult.Status.Success, "Duplicate Faces");
+        }
+
+        static ActionResult DuplicateFacesToObject()
+        {
+            int DuplicateedFaceCount = 0;
+            List<GameObject> Duplicateed = new List<GameObject>();
```
Comment (vkovec) [inline Editor/MenuActions/Geometry/DuplicateFaces.cs]: extra `e` in Duplicated in these variables

File: Tests/Editor/Geometry/DuplicateFacesTest.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                UObject.DestroyImmediate(selectables[i].gameObject);
+            }
+        }
+
+        [Test]
+        public void DuplicateFaces_ToObject()
+        {
+            //Make first faces selected
+            ProBuilderMesh mesh = selectables[0];
+            Assert.IsNotNull(mesh.faces);
```
Comment (karl-) [inline Tests/Editor/Geometry/DuplicateFacesTest.cs]: If this is just part of the test setup and not what you're actually testing, this should be `Assume.That(mesh.faces, Is.NotNull)`.

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+            foreach (ProBuilderMesh mesh in MeshSelection.topInternal)
+            {
+                if (mesh.selectedFaceCount < 1 || mesh.selectedFaceCount == mesh.facesInternal.Length)
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: This was removed later, indeed not required

File: Runtime/MeshOperations/ExtrudeElements.cs [hunk 4]
```diff
@@ -180,7 +180,7 @@ public static Edge[] Extrude(this ProBuilderMesh mesh, IEnumerable<Edge> edges,
         /// <param name="mesh">The source mesh.</param>
         /// <param name="faces">The faces to split from the mesh.</param>
         /// <returns>The faces created forming the detached face group.</returns>
-        public static List<Face> DetachFaces(this ProBuilderMesh mesh, IEnumerable<Face> faces)
+        public static List<Face> DetachFaces(this ProBuilderMesh mesh, IEnumerable<Face> faces, bool deleteSourceFaces = true)
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Runtime/MeshOperations/ExtrudeElements.cs]: will do noted.

**Comment by MartinImgSpc** [comment]
@karl-  do you want me to apply your review comment to DetachFaces action as this is almost an exact copy? Thanks!

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                mesh.Rebuild();
+                copy.Rebuild();
+
+                mesh.Optimize();
+                copy.Optimize();
+
+                mesh.ClearSelection();
+                copy.ClearSelection();
+
+                copy.gameObject.name = mesh.gameObject.name + "-Duplicate";
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: > After the duplicate action, the selection is lost. I think the selection should be set to the new duplicate faces.

I agree, I left it in line with the behavior of DetachFaces. Do you want me to update DetachFaces in the same way?

[the remark this answers is not in the document]
Comment (karl-) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: > Do you want me to update DetachFaces in the same way?

That would be great, thanks!

File: Editor/MenuActions/Geometry/DuplicateFaces.cs (hunk 1 above)
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: Done thanks!

File: Runtime/MeshOperations/ExtrudeElements.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Runtime/MeshOperations/ExtrudeElements.cs]: Done

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            if (count > 0)
+                return new ActionResult(ActionResult.Status.Success, "Duplicate " + count + (count > 1 ? " Faces" : " Face"));
+
+            return new ActionResult(ActionResult.Status.Success, "Duplicate Faces");
+        }
+
+        static ActionResult DuplicateFacesToObject()
+        {
+            int DuplicateedFaceCount = 0;
+            List<GameObject> Duplicateed = new List<GameObject>();
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: Done. Thank you!

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+            foreach (ProBuilderMesh mesh in MeshSelection.topInternal)
+            {
+                if (mesh.selectedFaceCount < 1 || mesh.selectedFaceCount == mesh.facesInternal.Length)
+                    continue;
+
+                var primary = mesh.selectedFaceIndexes.ToArray();
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: Thanks. Changed it in DetachFaces as well.

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                DuplicateedFaceCount += primary.Length;
+
+                List<int> inverse = new List<int>();
+
+                for (int i = 0; i < mesh.facesInternal.Length; i++)
+                    if (!primary.Contains(i))
+                        inverse.Add(i);
+
+                ProBuilderMesh copy = Object.Instantiate(mesh.gameObject, mesh.transform.parent).GetComponent<ProBuilderMesh>();
+                copy.MakeUnique();
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: Done.

File: Editor/MenuActions/Geometry/DuplicateFaces.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: Done!

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                copy.Optimize();
+
+                mesh.ClearSelection();
+                copy.ClearSelection();
+
+                copy.gameObject.name = mesh.gameObject.name + "-Duplicate";
+                Duplicateed.Add(copy.gameObject);
+            }
+
+            MeshSelection.SetSelection(Duplicateed.ToArray());
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/MenuActions/Geometry/DuplicateFaces.cs]: Thanks removed!

File: Tests/Editor/Geometry/DuplicateFacesTest.cs (hunk 2 above)
Comment (MartinImgSpc) [reply Tests/Editor/Geometry/DuplicateFacesTest.cs]: Done!

**Comment by gabrielw-us** [comment]
Made a few changes, committed and pushed - icon and tooltip edits. Then it should be good to go :)

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 10]
```diff
@@ -17,7 +17,7 @@ sealed class DuplicateFaces : MenuAction
         static readonly TooltipContent s_Tooltip = new TooltipContent
             (
                 "Duplicate Faces",
-                "Creates a new object (or submesh) from the selected faces."
+                "Makes an exact copy of the selected faces, and either adds them to this mesh or creates a new Game Object"
```
Comment (karl-) [inline Editor/MenuActions/Geometry/DuplicateFaces.cs]: @unitALG could you please review?

File: Editor/MenuActions/Geometry/DuplicateFaces.cs [hunk 11]
```diff
@@ -45,7 +45,7 @@ protected override void OnSettingsGUI()
         {
             GUILayout.Label("Duplicate Face Settings", EditorStyles.boldLabel);
 
-            EditorGUILayout.HelpBox("Duplicate Faces can separate the selection into either a new GameObject or a submesh.", MessageType.Info);
+            EditorGUILayout.HelpBox("You can create a new Game Object with the selected face(s), or keep them as part of this object by using a Submesh.", MessageType.Info);
```
Comment (karl-) [inline Editor/MenuActions/Geometry/DuplicateFaces.cs]: @unitALG could you please review?

**Pull request merged**